### ISSUE TITLE
DNS horizontal autoscaling tests failing during upgrade

### ISSUE THREAD

**Issue by ethernetdan** [kubernetes/kubernetes#43392]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG

**What happened**:
Tests for DNS horizontal autoscaling frequently fail in the upgrade suite. It seems 1.5 upgrades are affected much more than 1.4, possibly could be coincidence with such few tests run.
* [gce-1.5-1.6-upgrade-cluster-new
](https://k8s-testgrid.appspot.com/release-1.6-upgrade-skew#gce-1.5-1.6-upgrade-cluster-new)
* [gke-container_vm-1.5-container_vm-1.6-upgrade-cluster-new
](https://k8s-testgrid.appspot.com/release-1.6-upgrade-skew#gke-container_vm-1.5-container_vm-1.6-upgrade-cluster-new)
* [gke-container_vm-1.5-gci-1.6-upgrade-cluster-new](https://k8s-testgrid.appspot.com/release-1.6-upgrade-skew#gke-container_vm-1.5-gci-1.6-upgrade-cluster-new)
* [gke-gci-1.5-container_vm-1.6-upgrade-cluster-new
](https://k8s-testgrid.appspot.com/release-1.6-upgrade-skew#gke-gci-1.5-container_vm-1.6-upgrade-cluster-new)

**Comment by grodrigues3** [comment]
@ethernetdan are you working on this?  let me know if you need help assigning

**Comment by skriss** [comment]
cc @MrHohn who was looking at this test for #42766 - is this a different issue than the one you fixed last week?

**Comment by MrHohn** [comment]
Looking into it.

/assign

**Comment by MrHohn** [comment]
The serial test is hitting a similar issue ---  it expects to get the same replica count after letting dns_autoscaler to restore the default params but apparently the default params was changed after upgrade. So it will mostly fail. I will think of a way for fixing this.

Though the other test --- "kube-dns-autoscaler should scale kube-dns pods in both nonfaulty and faulty scenarios" --- is hitting a strange error:
```
Expected error:
    <*errors.errorString | 0xc4229c2890>: {
        s: "expected 1 autoscaler pod, got 2",
    }
    expected 1 autoscaler pod, got 2
not to have occurred
```
Seems like the old autoscaler did not got deleted after upgrade. I'm still looking into this one.

**Comment by ethernetdan** [comment]
@MrHohn any luck with a fix?

**Comment by MrHohn** [comment]
@ethernetdan Not yet :(

I checked couple CI posted above, and found some other add-ons (like kube-dashboard and l7-default-backend) were having similar situation --- the older version add-on didn't get cleaned up. I'm now working on reproducing this on my cluster.

**Comment by MrHohn** [comment]
Reproduced the issue after upgraded the master from 1.5 to 1.6 (nodes are still 1.5). Found all the add-ons deployments got updated correctly. But the underlying old pods did not get deleted by the deployments:
```
$ kubectl get pods --namespace=kube-system
NAME                                                 READY     STATUS    RESTARTS   AGE
etcd-empty-dir-cleanup-kubernetes-master             1/1       Running   0          17m
etcd-server-events-kubernetes-master                 1/1       Running   0          17m
etcd-server-kubernetes-master                        1/1       Running   0          17m
fluentd-cloud-logging-kubernetes-minion-group-95sh   1/1       Running   0          2h
fluentd-cloud-logging-kubernetes-minion-group-chxp   1/1       Running   0          2h
fluentd-cloud-logging-kubernetes-minion-group-cm98   1/1       Running   0          2h
fluentd-gcp-v2.0-zqdl4                               1/1       Running   0          17m
heapster-v1.3.0-beta.1-2540376499-gxwbc              4/4       Running   0          16m
kube-addon-manager-kubernetes-master                 1/1       Running   0          17m
kube-apiserver-kubernetes-master                     1/1       Running   0          17m
kube-controller-manager-kubernetes-master            1/1       Running   0          17m
kube-dns-4101612645-wrgsx                            4/4       Running   0          2h
kube-dns-autoscaler-1733107464-36c7g                 1/1       Running   0          17m
kube-dns-autoscaler-2715466192-v4r68                 1/1       Running   0          2h
kube-proxy-kubernetes-minion-group-95sh              1/1       Running   0          2h
kube-proxy-kubernetes-minion-group-chxp              1/1       Running   0          2h
kube-proxy-kubernetes-minion-group-cm98              1/1       Running   0          2h
kube-scheduler-kubernetes-master                     1/1       Running   0          17m
kubernetes-dashboard-2917723164-q39l7                1/1       Running   0          17m
kubernetes-dashboard-3543765157-73qgn                1/1       Running   0          2h
l7-default-backend-1044750973-mgqcp                  1/1       Running   0          17m
l7-default-backend-2234341178-2l4l1                  1/1       Running   0          2h
l7-lb-controller-v0.9.1-kubernetes-master            1/1       Running   0          17m
monitoring-influxdb-grafana-v4-ldzcr                 2/2       Running   0          2h
npd-v0.3.0-alpha.1-1xkq4                             1/1       Running   0          17m
npd-v0.3.0-alpha.1-dnf5g                             1/1       Running   0          17m
npd-v0.3.0-alpha.1-sbrc1                             1/1       Running   0          17m
rescheduler-v0.2.2-kubernetes-master                 1/1       Running   0          16m

$ kubectl get deploy --namespace=kube-system
NAME                     DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
heapster-v1.3.0-beta.1   1         1         1            1           23m
kube-dns                 1         1         1            1           2h
kube-dns-autoscaler      1         1         1            1           2h
kubernetes-dashboard     1         1         1            1           2h
l7-default-backend       1         1         1            1           2h
```

Need helps from @kubernetes/sig-apps-bugs

**Comment by MrHohn** [comment]
Also found couple errors in kube-controller-manager.log from https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-1.5-1.6-upgrade-cluster-new/32:
```
I0319 10:33:06.931909       5 deployment_controller.go:480] Error syncing deployment kube-system/kubernetes-dashboard: Deployment.apps "kubernetes-dashboard" is invalid: status.availableReplicas: Invalid value: 1: cannot be greater than readyReplicas
I0319 10:33:06.931949       5 deployment_controller.go:548] Started syncing deployment "kube-system/heapster-v1.2.0.1" (2017-03-19 10:33:06.931944867 +0000 UTC)
I0319 10:33:06.934486       5 deployment_controller.go:550] Finished syncing deployment "kube-system/kube-dns-autoscaler" (676.719179ms)
I0319 10:33:06.934518       5 deployment_controller.go:480] Error syncing deployment kube-system/kube-dns-autoscaler: Deployment.apps "kube-dns-autoscaler" is invalid: status.availableReplicas: Invalid value: 1: cannot be greater than readyReplicas
I0319 10:33:06.934568       5 deployment_controller.go:548] Started syncing deployment "kube-system/kube-dns" (2017-03-19 10:33:06.934563869 +0000 UTC)
I0319 10:33:06.936555       5 deployment_controller.go:550] Finished syncing deployment "kube-system/l7-default-backend" (663.940303ms)
I0319 10:33:06.936587       5 deployment_controller.go:480] Error syncing deployment kube-system/l7-default-backend: Deployment.apps "l7-default-backend" is invalid: status.availableReplicas: Invalid value: 1: cannot be greater than readyReplicas
I0319 10:33:06.936617       5 deployment_controller.go:548] Started syncing deployment "kube-system/kubernetes-dashboard" (2017-03-19 10:33:06.936613386 +0000 UTC)
...
I0319 10:33:06.990072       5 deployment_controller.go:276] ReplicaSet heapster-v1.3.0-4045285091 updated.
I0319 10:33:07.017979       5 servicecontroller.go:325] Not persisting unchanged LoadBalancerStatus for service kube-system/kubernetes-dashboard to registry.
I0319 10:33:07.018045       5 servicecontroller.go:746] Finished syncing service "kube-system/kubernetes-dashboard" (102.469794ms)
...
I0319 10:33:07.032320       5 replica_set.go:529] Finished syncing replica set "kube-system/heapster-v1.3.0-4045285091" (296.084028ms)
I0319 10:33:07.032843       5 replica_set_utils.go:71] Updating status for ReplicaSet: kube-system/heapster-v1.3.0-4045285091, replicas 0->1 (need 1), fullyLabeledReplicas 0->1, readyReplicas 0->0, availableReplicas 0->0, sequence No: 1->1
I0319 10:33:07.040245       5 deployment_controller.go:550] Finished syncing deployment "kube-system/kubernetes-dashboard" (103.606204ms)
I0319 10:33:07.041533       5 deployment_controller.go:480] Error syncing deployment kube-system/kubernetes-dashboard: Deployment.apps "kubernetes-dashboard" is invalid: status.availableReplicas: Invalid value: 1: cannot be greater than readyReplicas
I0319 10:33:07.041578       5 deployment_controller.go:548] Started syncing deployment "kube-system/kube-dns-autoscaler" (2017-03-19 10:33:07.041573788 +0000 UTC)
I0319 10:33:07.044714       5 deployment_controller.go:550] Finished syncing deployment "kube-system/heapster-v1.3.0" (771.531315ms)
I0319 10:33:07.044748       5 deployment_controller.go:480] Error syncing deployment kube-system/heapster-v1.3.0: Operation cannot be fulfilled on deployments.extensions "heapster-v1.3.0": the object has been modified; please apply your changes to the latest version and try again
```

**Comment by kargakis** [comment]
@MrHohn can you post the output of 
```
kubectl get deploy,rs,pods -l (labels from a component that had old pods) -o yaml
```
I wouldn't worry about the errors in the cm log because those deployments are getting retried on the spot but I personally worry about those errors because I am not sure how we can end up in such a (albeit transient) state.

**Comment by MrHohn** [comment]
Sure, seems like there are multiple replica sets existing. Please see the output on https://gist.github.com/MrHohn/294a79821322430a8e93e935f2b46de3.

**Comment by kargakis** [comment]
Can you grep for any "Dropping deployment" logs from the controller manager?

**Comment by MrHohn** [comment]
Found many:
```
I0319 10:34:08.125630       5 deployment_controller.go:486] Dropping deployment "kube-system/kube-dns-autoscaler" out of the queue: Deployment.apps "kube-dns-autoscaler" is invalid: status.availableReplicas: Invalid value: 1: cannot be greater than readyReplicas
...
I0319 10:35:45.178894       5 deployment_controller.go:486] Dropping deployment "kube-system/kubernetes-dashboard" out of the queue: Deployment.apps "kubernetes-dashboard" is invalid: status.availableReplicas: Invalid value: 1: cannot be greater than readyReplicas
...
I0319 10:35:45.185985       5 deployment_controller.go:486] Dropping deployment "kube-system/l7-default-backend" out of the queue: Deployment.apps "l7-default-backend" is invalid: status.availableReplicas: Invalid value: 1: cannot be greater than readyReplicas
...
I0319 10:36:39.683799       5 deployment_controller.go:486] Dropping deployment "kube-system/kube-dns" out of the queue: Deployment.apps "kube-dns" is invalid: status.availableReplicas: Invalid value: 1: cannot be greater than readyReplicas
```

**Comment by kargakis** [comment]
Ok, I just found the log.
```
I0319 10:33:07.765466       5 deployment_controller.go:486] Dropping deployment "kube-system/l7-default-backend" out of the queue: Deployment.apps "l7-default-backend" is invalid: status.availableReplicas: Invalid value: 1: cannot be greater than readyReplicas
```

**Comment by kargakis** [comment]
Ok, this problem shouldn't happen and my wild guess is that readyReplicas for Deployments is somehow dropped by protobuf due to the transition to the new api group.

**Comment by kargakis** [comment]
Looking further.... cc: @soltysh

**Comment by kargakis** [comment]
I think the issue stems from the fact that there is no ReadyReplicas field in 1.5 for Deployments. @MrHohn is the error happening before or after the update to 1.6? It seems that 1.6 validation is used for the 1.5 cluster.

**Comment by MrHohn** [comment]
It happened after the the master was updated to 1.6, though the nodes are still in v1.5.

**Comment by kargakis** [comment]
Yeah, the Deployment is in the apps group. Nodes shouldn't matter. The problem is that because there is no field in 1.5, when we update to 1.6 and the controller tries to update the Deployment, it will be denied by validation because the pre-existing availableReplicas field is greater than readyReplicas (normally readyReplicas should always be greater or equal). The fix for now is to disable the validation for readyReplicas. Opening a PR soon.

**Comment by MrHohn** [comment]
Status update in case it is not clear. The fix (#43465) sent by @kargakis will fix this issue, and the "old add-ons didn't get deleted" issue as well . And it is already LGTMed and Approved.
